### PR TITLE
#4061: Add moreh_bmm and moreh_linear ops

### PR DESCRIPTION
**Pull request by dongjin-na**
This PR implements moreh_bmm, moreh_linear ops. (https://github.com/tenstorrent-metal/tt-metal/issues/4060)

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/moreh_bmm/moreh_bmm_op.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "tensor/tensor.hpp"
+#include "tt_dnn/op_library/operation.hpp"
+
+namespace tt {
+namespace operations {
+namespace primary {
+
+using namespace tt_metal;
+
```
Comment (hschoi4448) [inline tt_eager/tt_dnn/op_library/moreh_bmm/moreh_bmm_op.hpp:17]: Why doesn't moreh_bmm have 'struct MorehBMM' like other operations?

Comment (dongjin-na) [reply tt_eager/tt_dnn/op_library/moreh_bmm/moreh_bmm_op.hpp:17]: moreh_bmm is a subset of moreh_matmul.
So internally it calls moreh_matmul op.

**Review by hschoi4448 — APPROVED** [review]

**Comment by jliangTT** [comment]
@jvasilje , @TT-BrianLiu @aliuTT  please refer to row 2 and 3 in https://docs.google.com/spreadsheets/d/1AaUz-MVB5H7eWXr3uncPjMOSqx0Ney0xbMKvY3tvBRY/edit#gid=0 for the test cases that Moreh did. Any specific input that we will require them to provide?

File: tt_eager/tt_dnn/op_library/moreh_linear_backward/kernels/moreh_bias_backward_multi_core_h.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+#include <cstdint>
+
+#include "compute_kernel_api/eltwise_binary.h"
+#include "compute_kernel_api/mask.h"
+#include "compute_kernel_api/reduce.h"
+#include "compute_kernel_api/tile_move_copy.h"
+
+ALWI void ACQ() { acquire_dst(tt::DstMode::Half); }
+ALWI void REL() { release_dst(tt::DstMode::Half); }
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/moreh_linear_backward/kernels/moreh_bias_backward_multi_core_h.cpp:14]: We are migrating to these new APIs:
- `acquire_dst` = `tile_regs_acquire` and `tile_regs_wait`
- `release_dst` = `tile_regs_commit` and `tile_regs_release`. 

Take a look at this file for documentation: `tt_metal/include/compute_kernel_api/reg_api.h`.
And this file for examples: `tt_eager/tt_dnn/op_library/bmm/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp`

It's fine for this PR, but I think it's worth switching over.

**Review by TT-BrianLiu — APPROVED** [review]

**Comment by TT-BrianLiu** [comment]
> @jvasilje , @TT-BrianLiu @aliuTT please refer to row 2 and 3 in https://docs.google.com/spreadsheets/d/1AaUz-MVB5H7eWXr3uncPjMOSqx0Ney0xbMKvY3tvBRY/edit#gid=0 for the test cases that Moreh did. Any specific input that we will require them to provide?

I think test cases are reasonable for these implementations

**Review by arakhmati — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/moreh_linear_backward/kernels/moreh_bias_backward_multi_core_h.cpp (hunk 2 above)
Comment (dongjin-na) [reply tt_eager/tt_dnn/op_library/moreh_linear_backward/kernels/moreh_bias_backward_multi_core_h.cpp:14]: Thank you for your suggestion. I'll apply it to my next PR. :)

**Pull request merged**